src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-                char* end{};
-                #endif
-
-                /* Position indices */
-                positionIndices.push_back(
-                    #ifndef CORRADE_TARGET_ANDROID
-                    std::stoul(indices[0])
-                    #else
-                    std::strtoul(indices[0].data(), &end, 10)
-                    #endif
-                    - positionIndexOffset);
-
-                #ifdef CORRADE_TARGET_ANDROID
-                if(end == indices[0].data()) throw std::exception{};
-                #endif
-
-                /* Texture coordinates */
-                if(indices.size() == 2 || (indices.size() == 3 && !indices[1].empty())) {
-                    textureCoordinateIndices.push_back(
-                        #ifndef CORRADE_TARGET_ANDROID
-                        std::stoul(indices[1])
-                        #else
-                        std::strtoul(indices[1].data(), &end, 10)
-                        #endif
-                        - textureCoordinateIndexOffset);
-
-                    #ifdef CORRADE_TARGET_ANDROID
-                    if(end == indices[1].data()) throw std::exception{};
-                    #endif
-                }
+UnsignedInt ObjImporter::doObject3DCount() const { return _state->meshes.size() + _state->meshlessObjects.size(); }
 
-                /* Normal indices */
-                if(indices.size() == 3) {
-                    normalIndices.push_back(
-                        #ifndef CORRADE_TARGET_ANDROID
-                        std::stoul(indices[2])
-                        #else
-                        std::strtoul(indices[2].data(), &end, 10)
-                        #endif
-                        - normalIndexOffset);
-
-                    #ifdef CORRADE_TARGET_ANDROID
-                    if(end == indices[2].data()) throw std::exception{};
-                    #endif
-                }
-            }
+Int ObjImporter::doObject3DForName(const std::string& name) {
+    auto result = _state->meshlessObjectIds.find(name);
+    if(result != _state->meshlessObjectIds.end()) {
+        return result->second;
+    }
 
-        /* Ignore unsupported keywords, error out on unknown keywords */
-        } else if(![&keyword](){
-            /* Using lambda to emulate for-else construct like in Python */
-            for(const std::string expected: {"mtllib", "usemtl", "g", "s"})
-                if(keyword == expected) return true;
-            return false;
-        }()) {
-            Error() << "Trade::ObjImporter::mesh3D(): unknown keyword" << keyword;
-            return std::nullopt;
+    result = _state->meshIds.find(name);
+    if(result != _state->meshIds.end()) {
+        return result->second;
+    }
+
+    return -1;
+}
+
+std::string ObjImporter::doObject3DName(UnsignedInt id) {
+    const size_t numMeshes = _state->meshes.size();
+    if(id >= numMeshes) {
+        return _state->meshlessObjects[id - numMeshes];
+    }
```

Comment:
Names for objects vs names for objects with meshes are retrieved differently, because for objects with meshes, the name is generated from the intermediate `ObjMesh` structure, dependent on primitive, group and object name and whether there are other primitives in the same group and other groups etc, for potentially ommiting those from the mesh/object name.